rbs/MethodTypeTranslator.cc

```diff
@@ -0,0 +1,307 @@
+        if (auto e = ctx.beginError(errLoc, core::errors::Rewriter::RBSError)) {
+            e.setHeader("Unexpected node type `{}` in method signature, expected `{}`", rbs_node_type_name(node->type),
+                        "Function");
+        }
+
+        return ast::MK::Untyped(methodType.loc);
+    }
+
+    // Collect type parameters
+
+    std::vector<std::pair<core::LocOffsets, core::NameRef>> typeParams;
+    for (rbs_node_list_node_t *list_node = node->type_params->head; list_node != nullptr; list_node = list_node->next) {
+        auto loc = TypeTranslator::nodeLoc(methodType.loc, list_node->node);
+
+        if (list_node->node->type != RBS_AST_TYPEPARAM) {
+            if (auto e = ctx.beginError(loc, core::errors::Rewriter::RBSError)) {
+                e.setHeader("Unexpected node type `{}` in type parameter list, expected `{}`",
+                            rbs_node_type_name(list_node->node), "TypeParam");
+            }
+
+            continue;
+        }
+
+        auto node = (rbs_ast_typeparam_t *)list_node->node;
+        rbs_constant_t *constant = rbs_constant_pool_id_to_constant(fake_constant_pool, node->name->constant_id);
+        std::string string(constant->start, constant->length);
+        typeParams.emplace_back(loc, ctx.state.enterNameUTF8(string));
+    }
+
+    // Collect positionals
+
+    rbs_types_function_t *functionType = (rbs_types_function_t *)node->type;
+    std::vector<RBSArg> args;
+
+    collectArgs(ctx, methodType.loc, functionType->required_positionals, args, false);
+
+    rbs_node_list_t *optionalPositionals = functionType->optional_positionals;
+    if (optionalPositionals && optionalPositionals->length > 0) {
+        collectArgs(ctx, methodType.loc, optionalPositionals, args, false);
+    }
+
+    rbs_node_t *restPositionals = functionType->rest_positionals;
+    if (restPositionals) {
+        if (restPositionals->type != RBS_TYPES_FUNCTION_PARAM) {
+            if (auto e = ctx.beginError(methodType.loc, core::errors::Rewriter::RBSError)) {
+                e.setHeader("Unexpected node type `{}` in rest positional argument, expected `{}`",
+                            rbs_node_type_name(restPositionals), "FunctionParam");
+            }
+        } else {
+            auto loc = TypeTranslator::nodeLoc(methodType.loc, restPositionals);
+            auto node = (rbs_types_function_param_t *)restPositionals;
+            auto arg = RBSArg{loc, node->name, node->type, false};
+            args.emplace_back(arg);
+        }
+    }
+
+    rbs_node_list_t *trailingPositionals = functionType->trailing_positionals;
+    if (trailingPositionals && trailingPositionals->length > 0) {
+        collectArgs(ctx, methodType.loc, trailingPositionals, args, false);
+    }
+
+    // Collect keywords
+
+    collectKeywords(ctx, methodType.loc, functionType->required_keywords, args, false);
+    collectKeywords(ctx, methodType.loc, functionType->optional_keywords, args, false);
```

Comment:
Is the second call really supposed to use `false` for `optional`?  If so, we never pass `true` to `collectKeywords`, so maybe we can delete the `optional` argument?  If it's supposed to use `true`, can we add test coverage that would have caught this case?